### ISSUE TITLE
Avoiding marking headers as links in VSCode

### ISSUE THREAD

**Issue by ilya-biryukov** [clangd/clangd#630]
Hi,

The clangd adds underscores to all includes in VSCode (see the screenshots below). I personally find this rather distracting and have been thinking of adding an option to disable this behavior. 
VSCode adds underlines based on the `documentLink` responses, so my initial thought is to avoid returning the results altogether.

I see a few options to do this:
1. implement in clangd C++ code, add an option to the clangd binary (a command-line flag?)
   + pros: probably the easiest to implement, can easily affect only includes if other kinds of links get added later, can be used in other editors if they misbehave in a similar manner.
   + cons: might be something other extensions don't care about.
2. implement in the VSCode extension. 
   + pros: clangd code unaffected.
   + cons: probably harder to implement (I'm worried this might require adding the middleware in the procotol communication), probably hard to filter other kinds of links if they ever get added.
3. remove `documentLink` from clangd altogether. Note that go-to-definition on the include enables navigation on ctrl+click and I don't see any added benefits.

I would personally prefer (3), and probably (1) if folks think we should keep it.
@sam-mccall, any thoughts on this?

Here are the screenshots with clangd on and off.
With clangd
![clangd](https://user-images.githubusercontent.com/809452/102687777-ae812880-4202-11eb-87ee-011445c7ab40.png)

Without clangd:
![no_clangd](https://user-images.githubusercontent.com/809452/102687778-afb25580-4202-11eb-9198-4e489ced5f1f.png)

**Comment by njames93** [comment]
I disagree with removing this outright, include files are a perfect use case for documentLink with internal links. I'd be leaning on the side of (1). 

The point of no added benefits I disagree with. Having a hover tip directing you to the include file is much more intuitive for a user (especially one who doesn't know the ins and outs of vscode/lsp/clangd). Where `as go to def` going to the include file, while convenient, is definitely not as intuitive. Also given that vscode offers `go to def` literally everywhere, even if you are hovering over white space, would likely result in a user just looking past go-to-def for navigating to a header file.

I do however see the frustration with the underline, would be nice if there was a way for vscode to disable that, but keep the hover tip. 
There is the `editor.links` config option. However that will disable all underlines and follow link hover tips, even for urls in comments which appear all over the place.

**Comment by ilya-biryukov** [comment]
There is `textDocument/hover` method for the hover tip. If we agree there's no added benefit apart from the hover tip, I think the way to go is to
(4) remove the include from `textDocument/documenLinks` and produce them using `textDocument/hover` instead.

**Comment by kirillbobyrev** [comment]
While I can understand your point, I think having underlines by default is quite valuable: I myself realized I can ctrl+click on header files to navigate them. Thus, I think one benefit of having those underlines is simply feature discovery. Sure, many users know about go-to-def on file links but I'm personally not sure how many of them know it when first trying the extension.

I got used to the UI and never really found it annoying but I can see why this can be somewhat distracting. Maybe optionally disabling this in some way would be a good option.

**Comment by njames93** [comment]
> There is `textDocument/hover` method for the hover tip. If we agree there's no added benefit apart from the hover tip, I think the way to go is to
> (4) remove the includes from `textDocument/documenLinks` and produce them using `textDocument/hover` instead.

That's not a great plan as it only makes sense when using vscode.

**Comment by ilya-biryukov** [comment]
@kirillbobyrev I don't think this was ever done for the sake of feature discovery, the underlines are just a by-product of using this particular LSP method.
Holding ctrl and moving the mouse around will show the indication that go-to is available, this is in line with how variables behave.
Would you put underlines under each variable references because ctrl+click allows you to go to that variable?

@njames93 what are the downsides for other editors? you only mentioned the hover tips, I believe this should be fully addressed in any editor that supports them.

**Comment by rapgenic** [comment]
I'd find the underlines quite distracting, too... And I think they make the text harder to read especially when they're combined with error squiggles:

![Schermata da 2020-12-23 16-03-04](https://user-images.githubusercontent.com/10487071/103010352-7a767200-4538-11eb-929c-f79ebfc0afe8.png)

Anyway, given that this seemed a simple task and that it's quite useful for me too, I tried to make a quick patch see:

https://reviews.llvm.org/D93763

**Comment by ilya-biryukov** [comment]
@rapgenic thanks for the implementation. Can't wait for it to land :)

**Comment by Delgan** [comment]
Sorry to bump but what's missing for the @rapgenic's Pull Request to be validated?

@sam-mccall Is there anything else to address regarding your comment: https://reviews.llvm.org/D93763#2479116?
